src/clj/harja/palvelin/palvelut/laadunseuranta/talvihoitoreitit_palvelu.clj

```diff
@@ -0,0 +1,187 @@
+                                                                               (tr/tr-osoite-moderni-fmt
+                                                                                 (:tie r) (:aosa r) (:aet r)
+                                                                                 (:losa r) (:let r)))})
+                                                              leikkaavat-geometriat)))
+                                                  [] (:sijainnit t)))
+
+                        ;; Tallennetaan mahdolliset virheet atomiin
+                        _ (swap! virheet-atom conj (conj leikkaavat-geometriat))
+
+                        ;; Jos talvihoitoreittiä ei löydy tietokannasta, niin tallennetaan se uutena
+                        talvihoitoreitti-id (when (and (nil? talvihoitoreitti-db) (empty? tieosoite-virheet) (empty? leikkaavat-geometriat))
+                                              (:id (talvihoitoreitit-q/lisaa-talvihoitoreitti-tietokantaan
+                                                     db t urakka-id (:id kayttaja))))
+
+                        ;; Jos talvihoitoreitti löytyy tietokannasta, niin päivitetään
+                        _ (when (and talvihoitoreitti-db (empty? tieosoite-virheet) (empty? leikkaavat-geometriat))
+                            (talvihoitoreitit-q/paivita-talvihoitoreitti-tietokantaan db t urakka-id (:id kayttaja))
+                            (swap! paivitetyt-atom conj (:reittinimi t)))]]
+
+            ;; Jos talvihoitoreitin perustiedot on onnistuneesti tallennettu, niin tallennetaan myös kalustot ja reitit
+            (when (and (nil? talvihoitoreitti-db) talvihoitoreitti-id)
+              (do
+                (talvihoitoreitit-q/lisaa-kalustot-ja-reitit db talvihoitoreitti-id t)
+                (swap! lisatyt-atom conj (:reittinimi t)))))
+
+        vastaus (if (and (empty? @lisatyt-atom) (empty? @virheet-atom) (empty? @paivitetyt-atom))
+                  {:virheet ["Excelistä ei löydetty talvihoitoreittejä."]}
+                  {:onnistuneet @lisatyt-atom
+                   :paivitetyt @paivitetyt-atom
+                   :virheet (flatten @virheet-atom)})]
+    ;; Tiedoston lukemisesta tullut request vaatii jostain syystä transit-vastauksen. JSON olisi default vaihtoehto,
+    ;; mutta koska transitiksi pakottaminen onnistuu, niin kutsujalle kelpaa JSONin kanssa myös transit.
+    (transit-vastaus vastaus)))
+
+(defn vastaanota-excel [db request]
+  (log/debug "vastaanota-excel :: request" request)
+  (oikeudet/vaadi-kirjoitusoikeus oikeudet/urakat-paikkaukset-paikkauskohteetkustannukset
+    (:kayttaja request)
+    (Integer/parseInt (get (:params request) "urakka-id")))
+  (let [urakka-id (Integer/parseInt (get (:params request) "urakka-id"))
+        kayttaja (:kayttaja request)]
+    ;; Tarkistetaan, että kutsussa on mukana urakka ja kayttaja
+    (if (and (not (nil? urakka-id))
```

Comment:
meniskö tää ihan vaan 
```clojure
(if (and urakka-id kayttaja)
``` 
eli tuokin tekee saman "not nil" tarkistuksen, mutta helpompi lukea.

Reply:
Meneehän se. 👍